javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,385 @@
+
+  var head: Node[E] = null
+  var last: Node[E] = null
+
+  var _size = 0
+
+  def getFirst(): E =
+    if (_size <= 0) throw new NoSuchElementException()
+    else peekFirst()
+
+  def getLast(): E =
+    if (_size <= 0) throw new NoSuchElementException()
+    else peekLast()
+
+  def removeFirst(): E = {
+    if (isEmpty()) throw new NoSuchElementException()
+    val ret = head.value
+
+    val oldHead = head
+    head = oldHead.next
+
+    if (head ne null) {
+      head.prev = null
+
+      if (head eq last) {
+        head.next = null
+        last = null
+      }
+    }
+
+    _size -= 1
+    ret
+  }
+
+  def removeLast(): E = {
+    if (isEmpty()) throw new NoSuchElementException()
+    if (last eq null) {
+      removeFirst()
+    } else {
+      val oldLast = last
+
+      if (oldLast.prev eq head) {
+        head.next = null
+        last = null
+      } else {
+        last = oldLast.prev
+        last.next = null
+      }
+      last = oldLast.prev
+      last.next = null
+
+      _size -= 1
+      oldLast.value
+    }
+  }
+
+  def addFirst(e: E): Unit = {
+    val oldHead = head
+
+    head = new Node(e, next = oldHead)
+
+    _size += 1
+
+    if (oldHead ne null) {
+      oldHead.prev = head
+
+      if ((last eq null) && (oldHead.next eq null))
+        last = oldHead
+    }
+  }
+
+  def addLast(e: E): Unit = {
+    if (head eq null) addFirst(e)
+    else {
+      val oldLast =
+        if (last ne null) last
+        else if (head ne null) head
+        else null
+
+      last = new Node(e, prev = oldLast)
+
+      _size += 1
+
+      if (oldLast ne null)
+        oldLast.next = last
+    }
+  }
+
+  override def contains(o: Any): Boolean =
+    iterator().exists(_ === o)
+
+  override def size(): Int =
+    _size
+
+  override def add(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  override def remove(o: Any): Boolean =
+    _removeOccurrence(listIterator, o)
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator
+    val changed = iter.hasNext()
+    while (iter.hasNext())
+      addLast(iter.next())
+
+    changed
+  }
+
+  override def clear(): Unit = {
+    head = null
+    last = null
+    _size = 0
+  }
+
+  private def getNodeAt(index: Int): Node[E] = {
+    if (index == 0) head
+    else if (index == (size - 1)) last
+    else {
+      var current: Node[E] = null
+      if (index <= (size/2)) {
+        current = head
+        for (_ <- 0 until index) current = current.next
+      } else {
+        current = last
+        for (_ <- index until (size - 1)) current = current.prev
+      }
+      current
+    }
+  }
+
+  override def get(index: Int): E = {
+    checkIndexInBounds(index)
+    getNodeAt(index).value
+  }
+
+  override def set(index: Int, element: E): E = {
+    checkIndexInBounds(index)
+    val node = getNodeAt(index)
+    val oldValue = node.value
+    node.value = element
+    oldValue
+  }
+
+  private def addNode(next: Node[E], e: E): Unit = {
```

Comment:
Could this be called `addBefore` or `addBeforeNode`? With just `addNode`, at the call site, it is unclear that you're adding *before* the node given as parameter.